erigon-lib/commitment/metrics.go

```diff
@@ -0,0 +1,338 @@
+		},
+	}
+}
+
+func (metrics *Metrics) Reset() {
+	metrics.Accounts.Reset()
+	metrics.Updates.Store(0)
+	metrics.AddressKeys.Store(0)
+	metrics.StorageKeys.Store(0)
+	metrics.LoadBranch.Store(0)
+	metrics.LoadAccount.Store(0)
+	metrics.LoadStorage.Store(0)
+	metrics.UpdateBranch.Store(0)
+	metrics.Unfolds.Store(0)
+	metrics.TotalUnfoldingTime = 0
+	metrics.TotalFoldingTime = 0
+	metrics.TotalProcessingTime = 0
+}
+
+func (metrics *Metrics) Now() time.Time {
+	if collectCommitmentMetrics {
+		return time.Now()
+	}
+	return time.Time{}
+}
+
+func (metrics *Metrics) TotalUnfoldingTimeInc(t time.Time) {
+	if collectCommitmentMetrics {
+		metrics.TotalUnfoldingTime += time.Since(t)
+	}
+}
+
+func (metrics *Metrics) TotalProcessingTimeInc(t time.Time) {
+	if collectCommitmentMetrics {
+		metrics.TotalProcessingTime += time.Since(t)
+	}
+}
+
+func (metrics *Metrics) TotalFoldingTimeInc(t time.Time) {
+	if collectCommitmentMetrics {
+		metrics.TotalFoldingTime += time.Since(t)
+	}
+}
+
+func NewAccounts() *AccountMetrics {
+	return &AccountMetrics{
+		AccountStats: make(map[string]*AccountStats),
+	}
+}
+
+type AccountStats struct {
+	AccountUpdates     uint64
+	StorageUpates      uint64
+	LoadBranch         uint64
+	LoadAccount        uint64
+	LoadStorage        uint64
+	Unfolds            uint64
+	TotalUnfoldingTime time.Duration
+	Folds              uint64
+	TotalFoldingTime   time.Duration
+}
+
+type AccountMetrics struct {
+	m sync.Mutex
+	// will be separate value for each key in parallel processing
+	currentPlainKey []byte
+	AccountStats    map[string]*AccountStats
+}
+
+func (processAccount *AccountMetrics) Headers() []string {
+	return []string{
+		"account",
+		"storage updates",
+		"loading branch",
+		"loading account",
+		"loading storage",
+		"total unfolds",
+		"total unfolding time (ms)",
+		"total folds",
+		"total folding time (ms)",
+	}
+}
+
+func (processAccount *AccountMetrics) Values() [][]string {
+	processAccount.m.Lock()
+	defer processAccount.m.Unlock()
+	values := make([][]string, len(processAccount.AccountStats), len(processAccount.AccountStats))
+	var ind int
+	for i, account := range processAccount.AccountStats {
+		values[ind] = []string{
+			i,
+			strconv.FormatUint(account.StorageUpates, 10),
+			strconv.FormatUint(account.LoadBranch, 10),
+			strconv.FormatUint(account.LoadAccount, 10),
+			strconv.FormatUint(account.LoadStorage, 10),
+			strconv.FormatUint(account.Unfolds, 10),
+			strconv.Itoa(int(account.TotalUnfoldingTime.Milliseconds())),
+			strconv.FormatUint(account.Folds, 10),
+			strconv.Itoa(int(account.TotalFoldingTime.Milliseconds())),
+		}
+		ind++
+	}
+	return values
+}
+
+func (processAccount *AccountMetrics) Reset() {
+	processAccount.m.Lock()
+	defer processAccount.m.Unlock()
+	processAccount.AccountStats = make(map[string]*AccountStats)
+}
+
+func (processAccount *AccountMetrics) UpdatesInc(plainKey []byte) {
+	if collectCommitmentMetrics && plainKey != nil {
+		processAccount.m.Lock()
```

Comment:
also could check if !collectMetrics and exit, reducing folding level of following code